logics/impls/controller/mod.rs

```diff
@@ -718,27 +694,239 @@ impl<T: Storage<Data>> Internal for T {
-        let caller_pool = if let Some(_pool_attribute) = pool_attribute {
-            Some((pool, _pool_attribute))
-        } else {
-            None
-        };
-        let (_, shortfall) = self._get_hypothetical_account_liquidity(
-            redeemer,
-            Some(pool),
-            redeem_amount,
-            0,
-            caller_pool,
-        )?;
-        if !shortfall.is_zero() {
-            return Err(Error::InsufficientLiquidity)
-        }
+        if let Some(oracle) = self._oracle() {
+            let mut asset_price: u128 = 0;
+            let mut liquidation_threshold: u128 = 0;
+            let (shortfall, account_data) = {
+                // For each asset the account is in
+                let caller = Self::env().caller();
+                let mut skip_pool = caller;
+
+                let markets = self._markets();
+                let mut asset_params = Vec::<HypotheticalAccountLiquidityCalculationParam>::new();
+
+                let mut total_collateral_in_base_currency = U256::from(0);
+                let mut avg_ltv = U256::from(0);
+                let mut avg_liquidation_threshold = U256::from(0);
+                let mut total_debt_in_base_currency: U256 = U256::from(0);
+
+                if let Some(pool_attribute) = pool_attributes {
+                    // if caller is a pool, get parameters for the pool without call the pool
+                    if pool_attribute.underlying.is_none() {
+                        return Err(Error::UnderlyingIsNotSet)
+                    }
+                    if pool_attribute.pool.is_none() {
+                        return Err(Error::PoolIsNotSet)
+                    }
+                    if pool_attribute.pool != Some(pool) {
+                        return Err(Error::PoolIsNotSet)
+                    }
 
-        // FEATURE: update governance token supply index & distribute
+                    let collateral_factor_mantissa: Option<WrappedU256> =
+                        self.collateral_factor_mantissa(pool_attribute.pool.unwrap());
+                    if collateral_factor_mantissa.is_none() {
+                        return Err(Error::MarketNotListed)
+                    }
+                    let ltv = U256::from(collateral_factor_mantissa.unwrap());
 
-        Ok(())
+                    liquidation_threshold = pool_attribute.liquidation_threshold;
+
+                    let oracle_price: Option<u128> =
+                        PriceOracleRef::get_price(&oracle, pool_attribute.underlying.unwrap());
+                    if let None | Some(0) = oracle_price {
+                        return Err(Error::PriceError)
+                    }
+                    let oracle_price_mantissa = Exp {
+                        mantissa: WrappedU256::from(U256::from(oracle_price.clone().unwrap())),
+                    };
+
+                    asset_price = oracle_price.unwrap();
+                    let compounded_liquidity_balance = pool_attribute.account_balance;
```

Comment:
It would be easier to see the following declarations considering the order of appearance

```rust
let compounded_liquidity_balance = pool_attribute.account_balance;
if compounded_liquidity_balance != 0 { ... }

let borrow_balance_stored = pool_attribute.account_borrow_balance;
if borrow_balance_stored != 0 { ... }
```